[Lee De En] iP

leedeen01 opened the pull request:
### MavisPro
>“Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes/)

MavisPro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/leedeen01/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [ ] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

#
If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

leedeen01 committed:
Rename chatbot name to "Mavis"

leedeen01 committed:
Add echoing feature to Mavis

leedeen01 committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested

leedeen01 committed:
Add the ability to mark tasks as done or undone

leedeen01 committed:
Add support for tracking three types of tasks using classes and inheritance

ToDos: tasks without any date/time attached to it
Deadlines: tasks that need to be done before a specific date/time
Events: tasks that start at a specific date/time and ends at a specific date/time

leedeen01 committed:
Label for level-4

leedeen01 committed:
Add test using the I/O redirection technique

leedeen01 committed:
Handle errors regarding invalid inputs

leedeen01 committed:
Add support for deleting tasks from the list

leedeen01 committed:
Change to ArrayList from Hashmap

leedeen01 committed:
Remove redundant line of code

leedeen01 committed:
Add javadocs for TaskManager, Task, Event, Deadline, and ToDo classes and methods

leedeen01 committed:
Update README.md

leedeen01 committed:
Update Javadoc for Mavis.java to include error handling details

leedeen01 committed:
Update javadocs on Task class functionality

leedeen01 committed:
Update javadocs on TaskManager class functionality

leedeen01 committed:
Mavis class: Add save feature

leedeen01 committed:
Merge branch 'branch-Level-7'

leedeen01 committed:
Standardize date and time formatting across codebase

Added consistent date and time formatting using DateTimeFormatter throughout the Deadline and Event classes for standardized parsing and display of dates

leedeen01 committed:
Merge branch-Level-8 into master

leedeen01 committed:
Mavis class: Update FILE_PATH to relative path

leedeen01 committed:
Refactor code to follow OOP principles

Extracted code into multiple classes for better organization and maintainability:
Ui: Handles user interactions and displays
Storage: Manages loading and saving tasks from/to a file
Parser: Interprets user commands and parses them
TaskList: Manages tasks, including adding and deleting tasks
Command: Manage and execute user command

leedeen01 committed:
Merge tag 'A-MoreOOP'

leedeen01 committed:
Divide classes into packages

leedeen01 committed:
Merge branch 'branch-A-Packages'

leedeen01 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

leedeen01 committed:
Add JUnit tests Ui, TaskList, Parser, Mavis and ToDo class

leedeen01 committed:
Add javadocs for all classes

leedeen01 committed:
Merge branch 'branch-A-JavaDoc'

leedeen01 committed:
Refactor code to comply with coding standards

leedeen01 committed:
Merge branch 'branch-A-CodingStandard'

leedeen01 committed:
Add find keyword command

leedeen01 committed:
Merge branch 'branch-A-Level-9'

leedeen01 committed:
Add checkstyle configuration and update code formatting

- Added checkstyle.xml and suppressions.xml to enforce project coding standards.
- Updated all source files to align with the new Checkstyle rules.
- Integrated Checkstyle into the build process for continuous code quality.

leedeen01 committed:
Merge branch 'branch-A-CheckStyle'

leedeen01 committed:
Add a JavaFX GUI to the chatbot

- Implement basic JavaFX UI with a text input field and response display area
- Create `MainWindow.fxml` and MainWindow class to define the main GUI layout
- Create `DialogBox.fxml` and DialogBog class to manage user and chatbot dialog display
- Add a separate `Launcher` class to handle JavaFX runtime issues

leedeen01 committed:
Merge branch 'branch-Level-10'

leedeen01 committed:
Use varargs in Parser for improved flexibility

- Replace fixed arguments with Java varargs where applicable
- Simplify method calls and improve code maintainability

leedeen01 committed:
Merge branch 'branch-A-Varargs'

Shamanbenny commented on src/main/java/mavis/Mavis.java:
```diff
@@ -0,0 +1,41 @@
+     * @param filePath The file path to load and save tasks.
+     */
+    public Mavis(final String filePath) {
+        ui = new Ui();
```
Would it be better to have `Ui` be a `static` class? (That way you won't have to instantiate `Ui` each time you require it)

Shamanbenny commented on src/main/java/mavis/Parser.java:
```diff
@@ -0,0 +1,126 @@
+                throw new MavisException("Invalid index for delete command.");
+            }
+            return new DeleteCommand(taskIndex);
+        } else if (input.startsWith("mark")) {
+            String[] parts = input.split(" ");
+            if (parts.length < 2) {
+                throw new MavisException("Mark command requires an index.");
+            }
+            int num;
+            try {
+                num = Integer.parseInt(parts[1]);
+            } catch (NumberFormatException e) {
+                throw new MavisException("Invalid index for mark command.");
+            }
+            return new MarkCommand(num);
+        } else if (input.startsWith("unmark")) {
+            String[] parts = input.split(" ");
+            if (parts.length < 2) {
+                throw new MavisException("Unmark command requires an index.");
+            }
+            int num;
+            try {
+                num = Integer.parseInt(parts[1]);
+            } catch (NumberFormatException e) {
+                throw new MavisException("Invalid index for unmark command.");
+            }
+            return new UnmarkCommand(num);
+        } else if (input.startsWith("find")) {
+            String toFind = input.substring(4).trim();
+            if (toFind.isEmpty()) {
+                throw new MavisException("Find command requires a search term.");
+            }
+            return new FindCommand(toFind);
+        } else if (input.equals("bye")) {
+            return new ExitCommand();
+        } else if (input.equals("list")) {
+            return new ListCommand();
+        } else {
+            throw new MavisException("I'm sorry, but I don't know what that means.");
+        }
```
I feel like it might be easier to grasp the entire structure of this if it was set in `switch case` instead?

Shamanbenny replied:
If I am also not mistaken, I believe `switch case` on String variable is done with Hash lookup, which results in an O(1) lookup rather than chaining `if else` statements which could result in O(n), where `n` is the number of conditions.
Read more about it [here](https://medium.com/@thedevelopernote/efficiency-of-if-vs-switch-in-java-a-quick-comparison-32f4ee0e20a3).

Shamanbenny commented on src/main/java/mavis/Parser.java:
```diff
@@ -0,0 +1,126 @@
+            }
+            String desc = parts[0].trim();
+            String byPart = parts[1].trim();
+            if (!byPart.toLowerCase().startsWith("by")) {
```
I like how you use `.startsWith()` to deal with the `/` commands. Quite an elegant and ideal solution!

Shamanbenny commented on src/main/java/mavis/Storage.java:
```diff
@@ -0,0 +1,109 @@
+    /**
+     * The file path to the Mavis data file for this task manager.
+     */
+    private String filePath;
```
Should this be `final` to prevent accidental assignment? If so, reminder to use `SCREAMING_SAKE_CASE` as referenced [here](https://se-education.org/guides/conventions/java/intermediate.html#naming).

Shamanbenny commented on src/main/java/mavis/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+
+/**
+ * The TaskList class provides functionality to manage a collection of tasks.
+ * It allows performing various operations on tasks, such as:
+ *
+ * - Adding new tasks
+ * - Deleting tasks by their identifier
+ * - Marking tasks as completed or incomplete
+ * - Listing all tasks with their current status
```
Would it be ideal to remove "lists" from the class header/JavaDoc? I believe it doesn't display as intended when hovering the class containing said JavaDoc on IDEs.

Shamanbenny commented on src/main/java/mavis/Storage.java:
```diff
@@ -0,0 +1,109 @@
+
+    /**
+     * The file path to the Mavis data file for this task manager.
+     */
```
Refer to the [Java coding standard on 'Comments'](https://se-education.org/guides/conventions/java/intermediate.html#comments), would it be better to use single line comments on class members?

Shamanbenny reviewed (commented):
Wonderful work thus far, aside from a few nits. Keep up the good coding standard!

leedeen01 committed:
Update runtest.sh to handle relative path

leedeen01 committed:
Add assertions to validate assumptions in Mavis, Parser and Storage

Assertions are not used to document key assumptions in Parser, Storage, and Mavis.

This makes it harder to catch unexpected errors and verify that critical components are initialized correctly.

Add assertions to ensure critical assumptions hold.
- In Parser, assert that input arrays are not null or empty before parsing.
- In Storage, assert that file paths are not null and parsed tasks are valid.
- In Mavis, assert that UI, Storage, and TaskList are properly initialized.

Assertions help catch logical errors early and improve code reliability without affecting runtime performance in production.

leedeen01 committed:
Refactor stage configuration in Main class

- Extracted stage configuration logic into configureStage method for better modularity.
- Set stage scene and adjusted minimum and maximum dimensions for a consistent UI experience.
- Improved code readability by separating concerns and making the code cleaner.

leedeen01 committed:
Merge pull request #1 from leedeen01/branch-A-Assertion

Add assertions to validate assumptions in Mavis, Parser and Storage

leedeen01 committed:
Merge pull request #2 from leedeen01/branch-A-CodeQuality

Refactor stage configuration in Main class

leedeen01 committed:
Add Javadoc comments for key classes

The class 'DialogBox', 'Launcher', 'Main', 'MainWindow', and 'Mavis'
lack Javadoc comments.

Adding Javadoc comments to these classes documents their
functionality and improve code clarity for future developers.

leedeen01 committed:
Detect and handle scheduling anomalies when adding tasks

- Implemented logic to detect if a new task clashes with existing tasks
  in the schedule.
- Checks for overlapping date ranges between tasks,
  ensuring no conflicts when adding new tasks.
- Throws an exception with a relevant message when a scheduling
  conflict is detected.

leedeen01 committed:
Refactor GUI to enhance asymmetry and add Pokémon theme

- Adjusted conversation layout to reflect the asymmetric nature of user-app interactions.
- Differentiated between user and app messages for better clarity.
- Polished overall design for improved user experience.
- Introduced a Pokémon theme to the interface, adding them

leedeen01 committed:
Enhance UI/UX to better align with theme

- Resized the profile picture to fit more dialog box

leedeen01 committed:
Refactor 'bye' handling

- Refresh the user profile picture on exit and reinitialization

leedeen01 committed:
Refactor Parser to use switch statements, and refine anomaly detection

- Replaced if-else with switch for cleaner command parsing
- Added Javadoc comments to Parser methods for better documentation
- Updated checkAnomalies to exclude self-comparison

leedeen01 committed:
Add a user guide

leedeen01 committed:
Fix errors in README

leedeen01 committed:
Improve Mavis User Guide formatting and clarity

leedeen01 committed:
Remove duplicate PNG file

leedeen01 committed:
Fix spacing issues in README

leedeen01 committed:
Fix README command formatting for better readability

leedeen01 committed:
Standardize command descriptions in README

leedeen01 committed:
Create 'data' directory and 'Mavis.txt' file if missing

leedeen01 committed:
Change 'done' to 'isDone' in Task

leedeen01 committed:
Edit Javadoc for methods in MainWindow and Launcher classes

leedeen01 committed:
Rename image to Ui.png
